### ISSUE TITLE
4.0 Roadmap

### ISSUE THREAD

**Issue by probablyup** [styled-components/styled-components#1694]
**planned**

- [x] ~~drop React 15 support and switch to React 16-compliant lifecycle methods, etc without polyfills~~
  - [x] ~~use React 16.3+ context API in ThemeProvider (https://github.com/styled-components/styled-components/pull/1894)~~
  - [x] ~~use `React.forwardRef` API to remove the need for `innerRef` (https://github.com/styled-components/styled-components/pull/1923)~~
- [x] ~~drop HTML element whitelist~~
  - [x] ~~desugar syntax via babel plugin (https://github.com/styled-components/babel-plugin-styled-components/pull/140)~~
  - [x] ~~add "no-tags" entry (https://github.com/styled-components/styled-components/pull/1912)~~
  - [x]  ~~replace imports to "styled-components" with "styled-components/no tags" entry when using the babel plugin (https://github.com/styled-components/babel-plugin-styled-components/pull/150)~~
- [x] ~~update documentation to strongly recommend using the babel plugin (https://github.com/styled-components/styled-components-website/pull/288)~~
- [x] ~~remove preprocess functionality (https://github.com/styled-components/babel-plugin-styled-components/pull/135)~~
- [x] ~~drop `.extend()` API (https://github.com/styled-components/styled-components/pull/1908)~~
- [x] ~~delete the now unnecessary `consolidateStreamedStyles()` API (https://github.com/styled-components/styled-components/pull/1906)~~
- [x] ~~component-based global style API (#1333)~~
  - [x] ~~remove the injectGlobal API (#1333)~~
- [x] ~~migrate TS typings out of the repo to DefinitelyTyped (https://github.com/styled-components/styled-components/issues/1778, https://github.com/DefinitelyTyped/DefinitelyTyped/pull/28207, https://github.com/styled-components/styled-components/pull/1922)~~
- [ ] add syntax highlighting support for new APIs (https://github.com/styled-components/webstorm-styled-components/issues/55, https://github.com/styled-components/stylelint-processor-styled-components/issues/209, https://github.com/styled-components/vscode-styled-components/issues/126, ~~https://github.com/styled-components/vim-styled-components/issues/48~~)
- [x] ~~drop prod builds of things and just let downstream webpack etc handle minification and such~~
- [x] ~~stop bundling PropTypes with the UMD build (https://github.com/styled-components/styled-components/pull/1921)~~
- [x] ~~document all new APIs, mark old APIs as deprecated in the docs~~
- [x] ~~remove jsnext:main from package.json (https://github.com/styled-components/styled-components/pull/1907)~~

**nice to have**

- generate sourcemaps (https://github.com/styled-components/styled-components/issues/827)
- investigate individual dynamic rule setting in production for perf (https://github.com/styled-components/styled-components/issues/779)
- drop HTML attribute regex + runtime checks
  - add reserved prop (name TBD) (https://github.com/styled-components/styled-components/pull/1682)
- config (via babel plugin?) for various things like disabling speedy mode, component namespacing, etc

**Comment by probablyup** [comment]
cc @mxstbr @kitten for commentary and suggestions

**Comment by mxstbr** [comment]
Global style component?

**Comment by probablyup** [comment]
Ah yes! Want to add it to the checklist?

**Comment by probablyup** [comment]
Did a tiny bit of hacking to see what the final size looks like after the first two checklist items:

before

<img width="578" alt="screen shot 2018-04-19 at 2 25 04 am" src="https://user-images.githubusercontent.com/570070/38974679-06fbd108-4379-11e8-900d-a8db98a7e167.png">

after

<img width="586" alt="screen shot 2018-04-19 at 2 25 07 am" src="https://user-images.githubusercontent.com/570070/38974680-0a41190e-4379-11e8-9a32-fd7be928b592.png">

😍 and that's before we even get into the React 16 optimizations

**Comment by probablyup** [comment]
@kitten I wasn't sure if I should add sweetsour to the checklist just yet? I'd love to target a mid-summer release for 4.0, depending on how much time we have.

**Comment by mxstbr** [comment]
@probablyup 😍😍😍😍😍😍

**Comment by probablyup** [comment]
@mxstbr I kinda see what you're saying but from the perspective of something using the library, I'd rather be told the correct way to do something from the onset so I didn't have to go back and refactor it all later, you know?

**Comment by mxstbr** [comment]
@probablyup but it's not a refactoring at all, it's fully automatic? Once you're ready to go full prod mode we recommend you drop in the Babel plugin—you just don't _have_ to use it from the onset.

**Comment by migueloller** [comment]
Are there any plans to change the `innerRef` prop to `ref` and leverage React 16.3 `forwardRef`?

**Comment by probablyup** [comment]
@migueloller yup definitely, that's part of the "drop React 15 support and switch to React 16-compliant lifecycle methods, etc without polyfills" item!

**Comment by probablyup** [comment]
https://github.com/styled-components/styled-components/tree/develop is the 4.0 branch 🎉

First big piece of this is going in soon, thanks to https://github.com/styled-components/styled-components/pull/1894

**Comment by alexandernanberg** [comment]
I think we can check off the "unsafe async-lifecycles" task after #1894 got merged. Can't find any references to `componentWillReceiveProps`, `componentWillMount` or `componentWillUpdate` anywhere but in the benchmark code (could send in a PR for this but the we might get misleading results compared to master)

I'd also like to add "Use `React.forwardRef` API" to the checklist, that way we won't have to use a `innerRef` prop anymore.

Stylish is having an ongoing rewrite so that's something we might want to add in v4 as well

**Comment by probablyup** [comment]
@alexandernanberg Yup I think that context API work ended up checking off a lot of things! 😄 

`forwardRef` is definitely something we want to do. You're welcome to pick it up if you have time!

**Comment by alexandernanberg** [comment]
@probablyup  Cool! Yeah I'd definitely like to do that when I got some extra time over 😄

**Comment by Andarist** [comment]
👋 From what I understand you are approaching v4 release and I would like to make a proposal for it. 

While prop whitelist is obviously non-ideal solution and has its costs (2kb+) the approach is used by more than 1 library, it would be great if we could just use for the time being the same module for it so the cost would be one time thing. 

I'd like to propose using [is-prop-valid](https://github.com/emotion-js/emotion/tree/master/next-packages/is-prop-valid). It's obviously used by emotion, but also by popmotion since recently and I know that if we could all agree on a single module @diegohaz would love to dedupe his copied `validAttr` from reakit.

Your solution is more correct (i.e. validating post `data-` characters) but IMHO this is not really needed, the solution without that extreme correctness is "good enough" and is also slightly lighter (~0.5kb from what I remember).

**Comment by probablyup** [comment]
@Andarist checking it out now

**Comment by probablyup** [comment]
@Andarist I think we can use it but they're missing some props - going to make an issue now

**Comment by Andarist** [comment]
Super! Just to add - I have not consulted any other emotion team member on this, just got the ball rolling here. I don't think anyone on the team will have anything against it though.

**Comment by ctretyak** [comment]
What about new typings for TypeScript like _createGlobalStyle_ ?

**Comment by Andarist** [comment]
The TS typings were moved out of the repository so that's job for DefinitelyTyped community, not styled-components team.

**Comment by Lavoaster** [comment]
@ctretyak I'm already on it. 😄  Watch [this](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29389) issue and then within a day after it gets merged it'll be available for use. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29389

**Comment by mxstbr** [comment]
Thanks so much @Lavoaster!
